Implement vehicle order table reversing

AaronVanGeffen opened the pull request:
Implements #1972

![Screenshot](https://github.com/OpenLoco/OpenLoco/assets/604665/2912689f-04f2-4da8-9ff4-48f782c62d9e) ![Screenshot (1)](https://github.com/OpenLoco/OpenLoco/assets/604665/68b3a2bd-37f9-4b7d-bb34-48f20386b5ef)

AaronVanGeffen commented on src/OpenLoco/src/Vehicles/Orders.cpp:
```diff
@@ -206,6 +206,68 @@ namespace OpenLoco::Vehicles
+        printf("End of table @ %p\n", dest);
+
+        // Figure out the new position of the order of interest
+        auto newOOIOffset = dest - orderOfInterest - kOrderSizes[enumValue(ooiType)];
+        printf("Calculated OOI @ %p\n", newOOIOffset);
```
This does not appear to be yielding the correct offset at present. Not sure what the problem is yet.

AaronVanGeffen commented on src/OpenLoco/src/Windows/Vehicle.cpp:
```diff
@@ -211,6 +212,7 @@ namespace OpenLoco::Ui::Windows::Vehicle
             makeWidget({ 240, 116 }, { 24, 24 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::route_delete, StringIds::tooltip_route_delete_order),
             makeWidget({ 240, 140 }, { 24, 12 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::red_arrow_up, StringIds::tooltip_route_move_order_up),
             makeWidget({ 240, 152 }, { 24, 12 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::red_arrow_down, StringIds::tooltip_route_move_order_down),
+            makeWidget({ 240, 164 }, { 24, 24 }, WidgetType::buttonWithImage, WindowColour::secondary, ImageIds::construction_right_turnaround),
```
Should add a tooltip for this new widget.

AaronVanGeffen commented on src/OpenLoco/src/Vehicles/Orders.cpp:
```diff
@@ -206,6 +206,45 @@ namespace OpenLoco::Vehicles
         return begin();
     }
 
+    uint16_t reverseVehicleOrderTable(uint16_t tableOffset, uint16_t orderOfInterest)
```
Probably should go into `OrderManager`, if we move `kOrderSizes` there as well.

LeftofZen reviewed (approved):
playtested this, seemed to work with no issues for me. it would be nice to preserve the ordering of the pickup-dropoff commands with destionations, but as discussed already that would require a lot more work, so this implementation is good for now

AaronVanGeffen commented:
@LeftofZen Thanks for testing. Yes, @YetiWiz's request is sensible, but a bunch more work with the current data structures.

Before merging this, I'd like to move `reverseVehicleOrderTable` to `OrderManager`. That becomes feasible after #1978 has been merged.

AaronVanGeffen commented:
Now that #1978 has been merged, I have rebased this PR and moved `reverseVehicleOrderTable` into `OrderManager`. This is now good to go, as far as I'm concerned.

duncanspumpkin commented on src/OpenLoco/src/Vehicles/OrderManager.cpp:
```diff
@@ -280,6 +280,45 @@ namespace OpenLoco::Vehicles::OrderManager
+    uint16_t reverseVehicleOrderTable(uint16_t tableOffset, uint16_t orderOfInterest)
+    {
+        // Retrieve list of raw orders
+        std::vector<uint64_t> rawOrders{};
```
We should probably have a note somewhere in the order manager that you can't use std:: algorithms for a lot of this due to the variable length of elements. Then hopefully when we stop using variable length we can simplify a huge amount of this.

AaronVanGeffen replied:
Completely agreed. I'll add the note in a few places. I'm thinking something among the lines of:
```
// Currently the use of std:: algorithms is not feasible due to variable order lengths
// TODO: simplify this after the changing the data structure for the order table
```

duncanspumpkin commented on src/OpenLoco/src/Windows/Vehicle.cpp:
```diff
@@ -193,12 +193,13 @@ namespace OpenLoco::Ui::Windows::Vehicle
         // 0x00500554
         static WindowEventList events;
-        constexpr uint64_t enabledWidgets = (1 << widx::routeList) | (1 << widx::orderForceUnload) | (1 << widx::orderWait) | (1 << widx::orderSkip) | (1 << widx::orderDelete) | (1 << widx::orderUp) | (1 << widx::orderDown) | Common::enabledWidgets;
+        constexpr uint64_t enabledWidgets = (1 << widx::routeList) | (1 << widx::orderForceUnload) | (1 << widx::orderWait) | (1 << widx::orderSkip) | (1 << widx::orderDelete) | (1 << widx::orderUp) | (1 << widx::orderDown) | (1 << widx::orderReverse) | Common::enabledWidgets;
```
should really be using 1ULL for constructing a uint64_t. I know we don't do that consistently through the codebase though.

duncanspumpkin commented on src/OpenLoco/src/Windows/Vehicle.cpp:
```diff
@@ -2480,6 +2482,18 @@ namespace OpenLoco::Ui::Windows::Vehicle
+            GameCommands::setErrorTitle(StringIds::empty);
+            auto result = GameCommands::doCommand(args, GameCommands::Flags::apply);
+
+            Vehicles::OrderManager::generateNumDisplayFrames(head); // Note: order changed, check if this matters.
```
note is not relevant for this

duncanspumpkin commented on data/language/en-GB.yml:
```diff
@@ -2334,3 +2334,4 @@ strings:
   2279: "{COLOUR BLACK}Checksum"
   2280: "Play sound effects for new messages"
   2281: "{SMALLFONT}{COLOUR BLACK}When enabled, new messages will be accompanied by a sound effect of an appropriately cheering crowd."
+  2282: "{SMALLFONT}{COLOUR BLACK}Reverse the route order table"
```
is it called order table in game?

AaronVanGeffen replied:
I hesitated to call it such as well. The tab is actually called 'Routing', but the command tooltips refer to 'orders'. As a compromise, I thought I'd refer to both in this way.
